Usage tracker web scale

colega opened the pull request:
This merges into the hackathon `usage-tracker` branch.

I would recommend checking each commit (and their description) individually to review faster and understand the order of thoughts, as multiple things changed here.

The main thing that is implemented here is the ownership of multiple partitions per usage-tracker instance, coordination of that ownership through Memberlist, graceful creation and shutdown of those partitions, and read-only mode that can be invoked through an HTTP endpoint just like it's done on the ingesters, intended to be performed by the rollout-operator annotation.

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -353,6 +353,25 @@ usage_tracker:
   [enabled: <boolean> | default = false]
 
+  # Number of partitions to use for the usage-tracker. This number isn't
+  # expected to change once usage-tracker is already being used.
```
```suggestion
  # expected to change after you're already using the usage-tracker.
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -353,6 +353,25 @@ usage_tracker:
+  # CLI flag: -usage-tracker.partition-reconcile-interval
+  [partition_reconcile_interval: <duration> | default = 10s]
+
+  # Time to wait beforeshutting down a partition that is no longer owned by this
```
```suggestion
  # Time to wait before shutting down a partition that is no longer owned by this
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -518,177 +533,175 @@ usage_tracker:
-      # Enable auto-creation of Kafka topic if it doesn't exist.
-      # CLI flag: -usage-tracker.events-storage.writer.auto-create-topic-enabled
+      # Enable auto-creation of Kafka topic on startup if it doesn't exist. If
+      # creating the topic fails and the topic doesn't already exist, Mimir will
+      # fail to start.
```
```suggestion
      # creating the topic fails and the topic doesn't already exist, Mimir
      # fails to start.
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -518,177 +533,175 @@ usage_tracker:
-      # auto-created topics could have an incorrect number of partitions.
-      # CLI flag: -usage-tracker.events-storage.writer.auto-create-topic-default-partitions
-      [auto_create_topic_default_partitions: <int> | default = 0]
+      # Mimir will create the topic with this number of partitions. When the
```
```suggestion
      # Mimir creates the topic with this number of partitions. When the
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -518,177 +533,175 @@ usage_tracker:
-      # CLI flag: -usage-tracker.events-storage.writer.auto-create-topic-default-partitions
-      [auto_create_topic_default_partitions: <int> | default = 0]
+      # Mimir will create the topic with this number of partitions. When the
+      # value is -1 the Kafka broker will use the default number of partitions
```
```suggestion
      # value is -1, the Kafka broker uses the default number of partitions
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -518,177 +533,175 @@ usage_tracker:
+      # The maximum number of concurrent fetch requests that the ingester makes
+      # when reading data from Kafka during startup. Concurrent fetch requests
+      # are issued only when there is sufficient backlog of records to consume.
+      # 0 to disable.
```
```suggestion
      # Set to 0 to disable.
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -697,138 +710,136 @@ usage_tracker:
-      # Enable auto-creation of Kafka topic if it doesn't exist.
-      # CLI flag: -usage-tracker.events-storage.reader.auto-create-topic-enabled
+      # Enable auto-creation of Kafka topic on startup if it doesn't exist. If
+      # creating the topic fails and the topic doesn't already exist, Mimir will
+      # fail to start.
```
```suggestion
      # creating the topic fails and the topic doesn't already exist, Mimir
      # fails to start.
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -697,138 +710,136 @@ usage_tracker:
-      # auto-created topics could have an incorrect number of partitions.
-      # CLI flag: -usage-tracker.events-storage.reader.auto-create-topic-default-partitions
-      [auto_create_topic_default_partitions: <int> | default = 0]
+      # Mimir will create the topic with this number of partitions. When the
```
```suggestion
      # Mimir creates the topic with this number of partitions. When the
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -697,138 +710,136 @@ usage_tracker:
-      # CLI flag: -usage-tracker.events-storage.reader.auto-create-topic-default-partitions
-      [auto_create_topic_default_partitions: <int> | default = 0]
+      # Mimir will create the topic with this number of partitions. When the
+      # value is -1 the Kafka broker will use the default number of partitions
```
```suggestion
      # value is -1, the Kafka broker uses the default number of partitions
```

tacole02 commented on docs/sources/mimir/configure/configuration-parameters/index.md:
```diff
@@ -697,138 +710,136 @@ usage_tracker:
+      # The maximum number of concurrent fetch requests that the ingester makes
+      # when reading data from Kafka during startup. Concurrent fetch requests
+      # are issued only when there is sufficient backlog of records to consume.
+      # 0 to disable.
```
```suggestion
      # Set to 0 to disable.
```

tacole02 reviewed (approved):
Thank you!

colega commented:
Thank you for your review, @tacole02 I addressed your comments in https://github.com/grafana/mimir/pull/10844/commits/d1f481af7c365b31690af013d8a0c394f1f264b7

Also, please note this is not intended to be merged to `main` yet (this is going to be merged into usage-tracker branch, which will have it's own review later)

pracucci commented on pkg/usagetracker/tracker.go:
```diff
@@ -152,195 +176,277 @@ func NewUsageTracker(cfg Config, partitionRing *ring.PartitionInstanceRing, over
+		level.Info(logger).Log("msg", "creating new partition")
+		p, err := newPartition(pid, t.cfg, t.partitionKVClient, t.partitionRing, t.eventsKafkaWriter, t.bucket, t, t.logger, t.registerer)
+		if err != nil {
+			return errors.Wrap(err, "unable to create partition")
```
[nit] When returning an error from this function, I would aways include the partition ID

colega replied:
Good point, done.

pracucci commented on pkg/usagetracker/tracker.go:
```diff
@@ -77,70 +98,73 @@ func (c *Config) Validate() error {
-	eventsKafkaReader *kgo.Client
 
-	pendingCreatedSeriesMarshaledEvents chan []byte
+	mtx        sync.RWMutex
```
[nit] I would rename it to `partitionsMtx` since it just protects `partitions`. We don't need to protect `lostPartitions` cause it's just used within `reconcilePartitions()` that is not called concurrently.

colega replied:
Done.

pracucci reviewed (approved):
Very nice work! I followed the logic quite easily 👏  I haven't reviewed all tiny details and I left many nits for myself, but overall makes sense to me.
